### ISSUE TITLE
Non-uniform shuffle in 3.1.0

### ISSUE THREAD

**Issue by tasuki** [elm-community/random-extra#24]
As of `3.1.0`, `Random.List.shuffle` produces non-uniform shuffle. Shuffling looks uniform(ish) when reverting to `3.0.0`. The relevant PRs are #14 and #17 (hi @owanturist!).

@miniBill created an Ellie, which I altered slightly to produce [20 independent rows, each with a `List.range 1 50` shuffled](https://ellie-app.com/7rK3JsCWSHma1). Here are 9 offenders from a single run, with added emphasis for certain regularities:

14 **7 6 5** 47 29 27 19 4 11 41 17 40 16 38 35 13 3 30 10 26 46 45 12 9 43 2 42 44 1 24 48 34 23 50 22 32 37 28 33 18 25 21 36 49 39 20 31 15 8
**11 9 8 5 4 3 1** 50 43 26 41 25 22 46 30 49 42 39 15 31 36 40 20 38 6 18 35 37 21 29 44 28 32 14 13 45 12 23 24 34 48 19 7 17 33 16 47 27 10 2
28 **44 43 41 40** 23 27 25 22 21 19 2 17 26 **16 15 14 12** 18 39 1 31 10 46 29 5 37 11 33 45 20 13 4 9 3 42 30 7 47 50 48 32 49 36 24 34 35 8 6 38
**48 45 44 43 42 41 40** 29 9 37 36 7 1 25 35 24 23 20 18 50 17 6 39 4 49 3 28 19 12 34 27 15 5 26 14 2 38 13 22 33 46 16 21 32 31 8 11 10 30 47
16 15 21 14 **9 8 5 4 2** 40 18 **39 38 36 34 33** 7 37 47 31 13 19 1 41 11 30 6 3 49 46 43 29 12 28 45 42 22 20 26 32 10 50 24 23 27 48 17 35 25 44
**44 43 42 39 35 34 33 32 28 26** 29 6 5 25 22 4 38 10 7 2 18 31 46 13 21 20 40 11 41 50 36 9 37 15 14 3 17 8 23 16 12 45 49 1 47 27 19 48 30 24
**49 48 47 45 44 39 38 37** 31 28 34 27 24 43 23 22 33 21 32 2 20 9 26 14 1 5 41 42 8 13 7 12 4 29 40 19 3 50 36 18 25 15 16 30 11 35 46 6 10 17
**36 35 34 33 32 31 29 28 26 25** 14 50 24 43 13 21 27 45 3 12 20 11 19 49 10 18 5 47 42 9 48 37 39 8 15 16 17 41 22 7 44 4 30 1 23 38 6 40 46 2
**20 18 17 16** 30 29 12 21 24 9 34 28 15 14 8 47 32 41 7 23 31 22 6 27 5 3 44 43 2 40 45 35 26 11 46 50 49 10 42 39 37 36 48 1 13 38 33 4 19 25

**Comment by tasuki** [comment]
Prior to encountering this, I didn't know anything about shuffling algorithms, and still know very very little. Tried to read up a bit, mostly everyone mentions the Fisher-Yates algorithm, which was used in `3.0.0` and abandoned in `3..1.0`. Does the current algorithm have a name?

From a cursory look at `src/Random/List.elm` and `src/Utils.elm`, several things jump out at me (thanks for commenting the code btw!):
```
-- It generates the sequence of random indexes
-- The indexes could and will (for sure) duplicate each other
-- But UnionFind will help us to convert them into uniq,
-- even if all of the indexes will be the same value
```

```
-- Finding corresponding leader of random index
-- If the index was already used the next one will be got
-- until not used index will be found
```

I'm not sure whether I really understand the algorithm, but this seems like it could create "cascades", leading to certain elements having a higher probability of being selected than others. It seems to me the use of `UnionFind` will always lead to some form of this (but hey I could be completely wrong).

**Comment by owanturist** [comment]
@tasuki thanks for the report!

I think you're completely right - the algorithm increase a probability of selecting a certain element comparing to others:
- assume we have a list of 4 elements to shuffle `[ 0, 1, 2, 3 ]`
- first of all we generate a sequence of random integers `[ 0, 4 )` (4 is length of the list)
- let's assume the first random index is `0`. Before selecting we had probabilities to select an element `[ 25%, 25%, 25%, 25% ]`. After selecting we transform the probabilities to `[ 0, 50%, 25%, 25% ]` because if we'll get another `0` index we select value from `the next` index (`1`) as well as if we'll get index `1` directly.

So that's definitely not random! I hope you didn't use the library to shuffle cards in casino app... 

To make it uniform we have to select a random element (not just next) in case when a target element is already selected. I think we can achieve that since we have a random number in hands. I'll make a PR this weekend.

**Comment by ianmackenzie** [comment]
If it's useful, I've taken the [alternative `shuffle` implementation](https://github.com/ianmackenzie/elm-fast-shuffle-experiment/blob/master/Main.elm) I mentioned in #9 and updated it to Elm 0.19. The basic algorithm is:

- Tag every list element with a random `Int`
- Call the built-in `List.sortBy` to sort by the random tag
- Strip off the tags to give a shuffled version of the original list

I _believe_ this method should produce uniform, non-biased shuffling, and in my tests (Chrome and Firefox on Windows 10) the `fastShuffle` function is 5-10x faster than `Random.Extra.shuffle`.

Right now I'm pretty focused on trying to finish up and publish my `elm-3d-scene` package, but I'm happy to answer questions on my implementation if @tasuki or anyone wants to use it to make a PR against `random-extra` =)

**Comment by owanturist** [comment]
@ianmackenzie that's nice and simple `O(n)` algorithm (current is `O(n logn)` and non uniform)! Could you make a PR to here?

**Comment by owanturist** [comment]
Oh, not `O(n)` since it uses sort but that's browsers' underhood fast solution so we can trust it.

**Comment by ianmackenzie** [comment]
Yeah I suspect a browser built-in O(n log n) is probably pretty comparable to a pure Elm O(n) in many cases =)

I'm probably going to be pretty busy over the next few weeks trying to finish up and publish `elm-3d-scene`, so I'm not sure how much time I'll have to put together a proper PR with documentation/tests etc. If I get some free time over the break I might be able to take a look, but in the meantime if someone else wants to take my code and turn it into a PR I'm happy to answer questions on the implementation.

**Comment by owanturist** [comment]
@ianmackenzie so I will be glad to implement your solution then and request you as reviewer if you don't mind.

**Comment by ianmackenzie** [comment]
@owanturist gladly, that sounds perfect!

**Comment by ianmackenzie** [comment]
@owanturist one thing to keep in mind: you might be tempted to move `decorate` and `undecorate` into a `let` block within `shuffle`, but I think the Elm compiler still has some issues where tail-recursive functions inside `let` blocks don't get properly optimized (turned into JavaScript `while` loops). So it's probably best to keep those as top-level functions, even if you end up renaming them to `decorateForShuffling` and `undecorateAfterShuffling` or something for clarity/to avoid potential naming conflicts.

**Comment by miniBill** [comment]
@ianmackenzie the generate-random-numbers, pair, order, unpair algorithm is almost perfect
Almost?
You're currently generating integers in the `[Random.minInt, Random.maxInt]` range, but for the sake of making the next example simpler let's suppose it's `[0,3]` (the reasoning still stands, you would just need huge lists for the problem to manifest).

Ok, suppose we use that method on a list of 200 items, every one will get paired with a random number (one of 0,1,2,3), then sorted.
We cannot assume that the algorithm that browsers use to sort is not stable, in particular we cannot assume that it will order items with the same decoration uniformly. If we could we would just decorate with a list of zeroes.
So, what is going to happen? The list will be split in four and then the fourths will be left unshuffled!
So, this algorithm is perfect as long as there are no duplicates in the list.

How can we fix this?
1. Switch to the old algorithm for huge lists (>2^20 elements I'd say)
2. Once you've generated the decorating list, remove dupes and then keep adding and checking for dupes until you reach the correct size

Step 2 will be almost free most of the time, because the collision probability will be low

**Comment by owanturist** [comment]
There is a [demo](https://ellie-app.com/7rTkZBqFSYza1) to play with the algorithm based on sort by random key. You could check how it distributes elements (if there are three+ items one by one they'll be marked in UI).

**Comment by tasuki** [comment]
Y'all people are wonderful! @ianmackenzie fast-shuffle is simple and easy to understand. @owanturist thanks for volunteering here! @miniBill great point wrt the maxInt limit!

> Once you've generated the decorating list, remove dupes and then keep adding and checking for dupes until you reach the correct size

I'm not sure I understand this though... keep adding and checking for dupes?

Let's suppose `[Random.minInt, Random.maxInt]` as `[0,3]`, then we split the list into 4 "buckets", if any bucket contains more than one element, the contents of the bucket need to be shuffled (and so on recursively turtles all the way down). Is that what you had in mind, or something else?

**Comment by miniBill** [comment]
The bucket idea might work too (and should still cost O(n log n)).

My idea was: take the random int list, remove dupes, add more random ints, removes dupes, ...

With regard to the cutoff point, the number of random ints to have >50% probability of a collision is actually about 46340, way lower than my initial guess. Yay birthday paradox.

This needs benchmarks!!!
Old method vs fast method with equality check vs fast method with buckets and turtles

**Comment by tasuki** [comment]
> My idea was: take the random int list, remove dupes, add more random ints, removes dupes

Oh I see. That works for `[Random.minInt, Random.maxInt]` being `[-2147483648, 2147483647]`, but wouldn't work for `[0,3]` and a list of 200 items, would it? I guess we'll never have a list of length anywhere close to 2^64, but still...

The bucket idea should work even if `[Random.minInt, Random.maxInt]` were `[0,3]`. The main disadvantage I can see is that it won't be tail recursive, so yay function calls? We probably won't blow the stack, but I think calling functions has a performance cost?

**Comment by ianmackenzie** [comment]
Great point @miniBill! A few thoughts:

- I think you could improve `fastShuffle` by using `Random.float 0 1` as the tag instead of `Random.int Random.minInt Random.maxInt`, which should give you a tag with 53 bits of randomness instead of 32. (I suspect in the original `fastShuffle` implementation that I was worried that using a random float instead of an int might lead to a non-uniform distribution of tags or something, but looking at [the `Random.float` implementation](https://github.com/elm/random/blob/ecf97bb43f0d5cd75243428f69f45323957bda25/src/Random.elm#L155) that doesn't look like it's the case.)
- For extremely large lists, you can probably still use decorate-sort-undecorate but decorate with a _list_ of `Int`s/`Float`s instead of a single one, with a heuristic for how large the list tag should be based on the size of the input list.

By my calculations, you can do up to ~100 million elements with 50% chance of tag collision (or up to ~10 million elements with a 1% chance) by using a 53-bit tag; in general, based on the exponential approximation from [here](https://en.wikipedia.org/wiki/Birthday_problem#Approximations), the number of bits of tag randomness you need for a given list size and probability of collision is

```elm
bits n p =
    ceiling (logBase 2 (toFloat n^2 / (-2 * logBase e (1 - p))))
```

So if you used a `List Float` as the random tag in the general case and wanted (for example) at most a 50% chance of collision, then the required size of the tag list could be calculated by something like

```elm
tagListSize n =
    ceiling (toFloat (bits n 0.5) / 53)
```

But given that I expect the vast majority of lists in Elm programs to be less than ten million elements, it's probably worth falling back to a `List Float` tag only for extremely large lists and using a single `Float` tag in all other cases. (I think most code could probably be shared between the two implementations by writing some generic functions parameterized on a `comparable` tag type...)

**Comment by ianmackenzie** [comment]
Actually a list tag might be overkill =) Even just using a `( Float, Float )` tag (106 bits of randomness) means that you could tag a list of ~10^15 (~2^50) elements with only a 1% chance of collision...and a list that large would require a few million gigabytes of RAM to store.

**Comment by ianmackenzie** [comment]
I wonder how much a small number of collisions are actually a problem...I think there's a good chance that just tagging with a `Float` (53 bits of randomness) will be enough for almost all use cases, and if you're shuffling a billion elements and really worried about randomness you can replace `Random.List.shuffle` with

```elm
Random.List.shuffle >> Random.andThen Random.List.shuffle
```

to just shuffle the list twice =)

**Comment by miniBill** [comment]
Yes, you are correct, 53 bits of random should be more than enough for normally-sized lists

**Comment by alexdarlington** [comment]
This bug really caught me out. Maybe the documentation could be updated to indicate there is a bug with shuffle and it doesn't produce a uniformly shuffled list?

**Comment by tasuki** [comment]
Ouch, @owanturist created #26 in _2019_. I wonder what needs to happen for that to move forward?

**Comment by owanturist** [comment]
Actually I don't know why it did stuck... @mgold does not have any activities for the last year, I hope everything is alright with him 🙏 

I don't know who else has permissions for merging.

**Comment by miniBill** [comment]
He does tweet [https://twitter.com/maxgoldst] so at least we know he exists still

**Comment by tasuki** [comment]
The fix was merged and [released in 3.2.0](https://github.com/elm-community/random-extra/releases/tag/3.2.0)! Thank you again everyone involved!
